File: implementations/java/client/src/test/java/com/theshuai/specusclient/peer/PeerServiceRuntimeTests.java
package com.theshuai.specusclient.peer;

import com.theshuai.common.clientauth.ClientAuthLoginResponse;
import com.theshuai.common.peermesh.LocalPeerService;
import com.theshuai.common.peermesh.PeerAdvertisedService;
import com.theshuai.common.peermesh.PeerControlMessage;
import com.theshuai.common.peermesh.PeerServiceDiscovery;
import com.theshuai.common.peermesh.PeerServiceSharingStatus;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import static org.assertj.core.api.Assertions.assertThat;

class PeerServiceRuntimeTests {
    private final List<String> sent = new CopyOnWriteArrayList<>();
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
    private PeerServiceRuntime runtime;

    @AfterEach
    void tearDown() {
        if (runtime != null) {
            runtime.close();
        }
        scheduler.shutdownNow();
    }

    @Test
    void doesNotProbeOrReportWhenSharingIsOff() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setHasAuthorizedOnlinePeer(true);
            runtime.applyConfig(config(false, localService(port, true)));
            Thread.sleep(50);
            assertThat(sent).isEmpty();
            assertThat(runtime.remoteServices()).isEmpty();
        }
    }

    @Test
    void doesNotProbeWithoutAuthorizedOnlinePeer() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setHasAuthorizedOnlinePeer(false);
            runtime.applyConfig(config(true, localService(port, true)));
            Thread.sleep(50);
            assertThat(sent).isEmpty();
        }
    }

    @Test
    void reportsReachableEnabledServiceAndBuildsSafeCatalogUrl() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setRosterLookup(id -> new PeerServiceRuntime.RosterHint("100.96.0.2", true));
            runtime.setHasAuthorizedOnlinePeer(true);
            runtime.applyConfig(config(true, localService(port, true)));
            waitUntil(() -> !sent.isEmpty());
            assertThat(sent.getFirst()).contains("service-report");
            assertThat(sent.getFirst()).contains("svc-http01");
            assertThat(sent.getFirst()).doesNotContain("targetHost");

            PeerControlMessage catalog = catalogMessage();
            runtime.applyCatalog(catalog);
            List<PeerServiceRuntime.RemoteServiceView> views = runtime.remoteServices();
            assertThat(views).hasSize(1);
            assertThat(views.getFirst().openable()).isTrue();
            assertThat(views.getFirst().accessTarget()).isEqualTo("http://100.96.0.2:8080/app");
            assertThat(views.getFirst().accessTarget()).doesNotContain("evil");
        }
    }

    @Test
    void localPauseStopsReportingWithoutChangingServerConfig() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setHasAuthorizedOnlinePeer(true);
            runtime.applyConfig(config(true, localService(port, true)));
            waitUntil(() -> !sent.isEmpty());
            sent.clear();
            runtime.setLocalPublished("svc-http01", false);
            waitUntil(() -> sent.stream().anyMatch(body -> body.contains("\"enabled\":false")
                    || body.contains("\"services\":[]")
                    || body.contains("\"services\": []")));
            assertThat(runtime.isLocallyPublished("svc-http01")).isFalse();
        }
    }

    @Test
    void turningSharingOffWithdrawsPreviousReport() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setHasAuthorizedOnlinePeer(true);
            runtime.applyConfig(config(true, localService(port, true)));
            waitUntil(() -> !sent.isEmpty());
            sent.clear();
            runtime.applyConfig(config(false, localService(port, true)));
            waitUntil(() -> sent.stream().anyMatch(body -> body.contains("\"enabled\":false")));
            assertThat(sent.getLast()).contains("service-report");
            assertThat(sent.getLast()).contains("\"enabled\":false");
        }
    }

    @Test
    void emptyCatalogWithdrawsAndOfflinePublisherDisablesOpen() {
        runtime = newRuntime();
        runtime.applyConfig(config(true, localService(1, false)));
        runtime.setRosterLookup(id -> new PeerServiceRuntime.RosterHint("100.96.0.2", false));
        runtime.applyCatalog(catalogMessage());
        assertThat(runtime.remoteServices().getFirst().openable()).isFalse();
        assertThat(runtime.remoteServices().getFirst().unavailableReason()).contains("离线");

        PeerControlMessage empty = catalogMessage();
        empty.setRevision(2L);
        empty.setServices(List.of());
        runtime.applyCatalog(empty);
        assertThat(runtime.remoteServices()).isEmpty();
    }

    @Test
    void staleCatalogCannotRollBackOrReviveWithdrawnServices() {
        runtime = newRuntime();
        runtime.applyConfig(config(true, localService(1, false)));
        PeerControlMessage newest = catalogMessage();
        newest.setRevision(2L);
        runtime.applyCatalog(newest);
        PeerControlMessage stale = catalogMessage();
        stale.setRevision(1L);
        runtime.applyCatalog(stale);
        assertThat(runtime.remoteServices()).hasSize(1);

        PeerControlMessage withdrawn = catalogMessage();
        withdrawn.setRevision(3L);
        withdrawn.setServices(List.of());
        runtime.applyCatalog(withdrawn);
        runtime.applyCatalog(newest);
        assertThat(runtime.remoteServices()).isEmpty();
    }

    @Test
    void reconnectClearsCatalogRevisionHighWaterAndAcceptsCurrentSnapshot() {
        runtime = newRuntime();
        runtime.applyConfig(config(true, localService(1, false)));
        runtime.setRosterLookup(id -> new PeerServiceRuntime.RosterHint("100.96.0.2", true));
        runtime.setHasAuthorizedOnlinePeer(true);
        PeerControlMessage current = catalogMessage();
        current.setRevision(7L);
        runtime.applyCatalog(current);
        assertThat(runtime.remoteServices()).hasSize(1);

        runtime.setHasAuthorizedOnlinePeer(false);
        assertThat(runtime.remoteServices()).isEmpty();
        runtime.setHasAuthorizedOnlinePeer(true);
        runtime.applyCatalog(current);

        assertThat(runtime.remoteServices()).hasSize(1);
    }

    @Test
    void unchangedServicesAreRenewedAcrossMultipleCatalogTtls() throws Exception {
        int port = freePort();
        try (ServerSocket ignored = listen(port)) {
            runtime = newRuntime();
            runtime.setHasAuthorizedOnlinePeer(true);
            runtime.applyConfig(config(true, localService(port, true)));
            waitUntil(() -> !sent.isEmpty());
            sent.clear();
            for (int elapsedTtls = 1; elapsedTtls <= 3; elapsedTtls++) {
                runtime.lastReportAt = Instant.now()
                        .minus(PeerServiceRuntime.REPORT_REFRESH_INTERVAL)
                        .minusSeconds(1);
                runtime.probeAndReport();
                assertThat(sent).hasSize(elapsedTtls);
            }
        }
    }

    @Test
    void probeTcpDetectsOpenAndClosedPorts() throws Exception {
        int port = freePort();
        assertThat(PeerServiceDiscovery.probeTcp("127.0.0.1", port, 200)).isFalse();
        try (ServerSocket ignored = listen(port)) {
            assertThat(PeerServiceDiscovery.probeTcp("127.0.0.1", port, 400)).isTrue();
        }
    }

    @Test
    void tcpBridgeRejectsSourceOutsideServerAuthoredAcl() throws Exception {
        int targetPort = freePort();
        int publishedPort = freePort();
        LocalPeerService local = localService(targetPort, true);
        local.setPublishedPort(publishedPort);
        local.setAllowedPeerVirtualIps(List.of("127.0.0.2"));
        try (ServerSocket target = listen(targetPort);
             PeerServiceBridge bridge = PeerServiceBridge.bind("127.0.0.1", local);
             Socket caller = new Socket("127.0.0.1", publishedPort)) {
            target.setSoTimeout(250);
            assertThat(org.assertj.core.api.Assertions.catchThrowable(target::accept))
                    .isInstanceOf(SocketTimeoutException.class);
            assertThat(caller.getInputStream().read()).isEqualTo(-1);
        }
    }

    @Test
    void tcpBridgeAllowsServerAuthorizedSource() throws Exception {
        int targetPort = freePort();
        int publishedPort = freePort();
        LocalPeerService local = localService(targetPort, true);
        local.setPublishedPort(publishedPort);
        local.setAllowedPeerVirtualIps(List.of("127.0.0.1"));
        try (ServerSocket target = listen(targetPort);
             PeerServiceBridge bridge = PeerServiceBridge.bind("127.0.0.1", local);
             Socket caller = new Socket("127.0.0.1", publishedPort);
             Socket forwarded = target.accept()) {
            assertThat(forwarded.isConnected()).isTrue();
        }
    }

    @Test
    void tcpBridgeSeparatesThreePeersAndRevocationClosesTheActiveFlow() throws Exception {
        int targetPort = freePort();
        int publishedPort = freePort();
        LocalPeerService local = localService(targetPort, true);
        local.setPublishedPort(publishedPort);
        local.setAllowedPeerVirtualIps(List.of("127.0.0.2"));
        try (ServerSocket target = listen(targetPort);
             PeerServiceBridge bridge = PeerServiceBridge.bind("127.0.0.1", local);
             Socket denied = connectFrom("127.0.0.3", publishedPort)) {
            target.setSoTimeout(250);
            assertThat(org.assertj.core.api.Assertions.catchThrowable(target::accept))
                    .isInstanceOf(SocketTimeoutException.class);

            try (Socket allowed = connectFrom("127.0.0.2", publishedPort)) {
                target.setSoTimeout(1_000);
                try (Socket forwarded = target.accept()) {
                    assertThat(forwarded.isConnected()).isTrue();
                    allowed.setSoTimeout(1_000);
                    bridge.close();
                    int result;
                    try {
                        result = allowed.getInputStream().read();
                    } catch (IOException closed) {
                        result = -1;
                    }
                    assertThat(result).isEqualTo(-1);
                }
            }
            assertThat(org.assertj.core.api.Assertions.catchThrowable(
                    () -> connectFrom("127.0.0.2", publishedPort)))
                    .isInstanceOf(IOException.class);
        }
    }

    @Test
    void udpBridgeAppliesTheSameAclAndRevocationBoundary() throws Exception {
        byte[] payload = {1, 2, 3};
        try (DatagramSocket target = new DatagramSocket(new InetSocketAddress("127.0.0.1", 0));
             DatagramSocket denied = new DatagramSocket(new InetSocketAddress("127.0.0.3", 0));
             DatagramSocket allowed = new DatagramSocket(new InetSocketAddress("127.0.0.2", 0))) {
            int publishedPort = freeUdpPort();
            LocalPeerService local = localService(target.getLocalPort(), true);
            local.setTransport("udp");
            local.setApplication("udp");
            local.setPublishedPort(publishedPort);
            local.setAllowedPeerVirtualIps(List.of("127.0.0.2"));
            try (PeerServiceUdpBridge bridge = PeerServiceUdpBridge.bind("127.0.0.1", local)) {
                denied.send(new DatagramPacket(payload, payload.length,
                        new InetSocketAddress("127.0.0.1", publishedPort)));
                target.setSoTimeout(250);
                assertThat(org.assertj.core.api.Assertions.catchThrowable(
                        () -> target.receive(new DatagramPacket(new byte[8], 8))))
                        .isInstanceOf(SocketTimeoutException.class);

                allowed.send(new DatagramPacket(payload, payload.length,
                        new InetSocketAddress("127.0.0.1", publishedPort)));
                target.setSoTimeout(1_000);
                DatagramPacket forwarded = new DatagramPacket(new byte[8], 8);
                target.receive(forwarded);
                assertThat(forwarded.getLength()).isEqualTo(payload.length);

                bridge.close();
                allowed.send(new DatagramPacket(payload, payload.length,
                        new InetSocketAddress("127.0.0.1", publishedPort)));
                target.setSoTimeout(250);
                assertThat(org.assertj.core.api.Assertions.catchThrowable(
                        () -> target.receive(new DatagramPacket(new byte[8], 8))))
                        .isInstanceOf(SocketTimeoutException.class);
            }
        }
    }

    private static Socket connectFrom(String sourceIp, int targetPort) throws IOException {
        Socket socket = new Socket();
        try {
            socket.bind(new InetSocketAddress(InetAddress.getByName(sourceIp), 0));
            socket.connect(new InetSocketAddress(InetAddress.getLoopbackAddress(), targetPort));
            return socket;
        } catch (IOException failure) {
            socket.close();
            throw failure;
        }
    }

    private static int freeUdpPort() throws IOException {
        try (DatagramSocket socket = new DatagramSocket(0)) {
            return socket.getLocalPort();
        }
    }

    private PeerServiceRuntime newRuntime() {
        return new PeerServiceRuntime((to, message) -> sent.add(message), scheduler);
    }

    private static ClientAuthLoginResponse.PeerMeshConfig config(boolean sharing, LocalPeerService local) {
        ClientAuthLoginResponse.PeerMeshConfig config = new ClientAuthLoginResponse.PeerMeshConfig();
        config.setEnabled(true);
        config.setVirtualIp("127.0.0.1");
        config.setServiceSharing(PeerServiceSharingStatus.of(true, sharing, true));
        config.setLocalServices(List.of(local));
        return config;
    }

    private static LocalPeerService localService(int targetPort, boolean enabled) {
        LocalPeerService local = new LocalPeerService();
        local.setServiceId("svc-http01");
        local.setName("web");
        local.setTransport("tcp");
        local.setApplication("http");
        local.setTargetHost("127.0.0.1");
        local.setTargetPort(targetPort);
        local.setPublishedPort(18080);
        local.setPath("/app");
        local.setEnabled(enabled);
        return local;
    }

    private static PeerControlMessage catalogMessage() {
        PeerAdvertisedService service = new PeerAdvertisedService();
        service.setServiceId("svc-http01");
        service.setName("web");
        service.setTransport("tcp");
        service.setApplication("http");
        service.setPublishedPort(8080);
        service.setPath("/app");
        PeerControlMessage catalog = new PeerControlMessage();
        catalog.setType(PeerControlMessage.TYPE_SERVICE_CATALOG);
        catalog.setPublisherClientId(2L);
        catalog.setPublisherClientName("client-b");
        catalog.setPublisherSessionId(9L);
        catalog.setRevision(1L);
        catalog.setExpiresAt(Instant.now().plusSeconds(60).toString());
        catalog.setServices(List.of(service));
        return catalog;
    }

    private static int freePort() throws Exception {
        try (ServerSocket socket = new ServerSocket(0)) {
            return socket.getLocalPort();
        }
    }

    private static ServerSocket listen(int port) throws Exception {
        ServerSocket socket = new ServerSocket(port);
        socket.setReuseAddress(true);
        return socket;
    }

    private static void waitUntil(java.util.concurrent.Callable<Boolean> condition) throws Exception {
        for (int i = 0; i < 50; i++) {
            if (Boolean.TRUE.equals(condition.call())) {
                return;
            }
            Thread.sleep(20);
        }
        throw new AssertionError("condition not met");
    }
}

File: implementations/csharp/client/tests/Specus.Client.Tests/PeerServiceRuntimeTests.cs
using System.Net;
using System.Net.Sockets;
using Specus.Client.Configuration;
using Specus.Client.PeerMesh;

namespace Specus.Client.Tests;

public class PeerServiceRuntimeTests
{
    [Fact]
    public void DoesNotProbeOrReportWhenSharingIsOff()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyConfig(Config(sharing: false, port, enabled: true));
        Thread.Sleep(50);
        Assert.Empty(sent);
    }

    [Fact]
    public void DoesNotProbeWithoutAuthorizedOnlinePeer()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetHasAuthorizedOnlinePeer(false);
        runtime.ApplyConfig(Config(sharing: true, port, enabled: true));
        Thread.Sleep(50);
        Assert.Empty(sent);
    }

    [Fact]
    public void ReportsReachableEnabledServiceAndBuildsSafeCatalogUrl()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetRosterLookup(_ => new PeerServiceRuntime.RosterHint("100.96.0.2", true));
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyConfig(Config(sharing: true, port, enabled: true));
        WaitUntil(() => sent.Count > 0);
        Assert.Contains("service-report", sent[0]);
        Assert.Contains("svc-http01", sent[0]);
        Assert.DoesNotContain("targetHost", sent[0]);

        runtime.ApplyCatalog(2, "client-b", 9, 1, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"),
        [
            new AdvertisedService
            {
                ServiceId = "svc-http01",
                Name = "web",
                Transport = "tcp",
                Application = "http",
                PublishedPort = 8080,
                Path = "/app",
            },
        ]);
        var views = runtime.RemoteServices();
        Assert.Single(views);
        Assert.True(views[0].Openable);
        Assert.Equal("http://100.96.0.2:8080/app", views[0].AccessTarget);
        Assert.DoesNotContain("evil", views[0].AccessTarget);
    }

    [Fact]
    public void EmptyCatalogWithdrawsAndOfflinePublisherDisablesOpen()
    {
        using var runtime = new PeerServiceRuntime(_ => { });
        runtime.SetRosterLookup(_ => new PeerServiceRuntime.RosterHint("100.96.0.2", false));
        runtime.ApplyConfig(Config(sharing: true, FreePort(), enabled: false));
        runtime.ApplyCatalog(2, "client-b", 9, 1, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"),
        [
            new AdvertisedService
            {
                ServiceId = "svc-http01",
                Name = "web",
                Transport = "tcp",
                Application = "http",
                PublishedPort = 8080,
                Path = "/app",
            },
        ]);
        Assert.False(runtime.RemoteServices()[0].Openable);
        Assert.Contains("离线", runtime.RemoteServices()[0].UnavailableReason);

        runtime.ApplyCatalog(2, "client-b", 9, 2, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), []);
        Assert.Empty(runtime.RemoteServices());
    }

    [Fact]
    public void CatalogRevisionTombstoneRejectsRollbackAndLateRevival()
    {
        using var runtime = new PeerServiceRuntime(_ => { });
        runtime.SetRosterLookup(_ => new PeerServiceRuntime.RosterHint("100.96.0.2", true));
        runtime.ApplyConfig(Config(sharing: true, FreePort(), enabled: false));
        runtime.SetHasAuthorizedOnlinePeer(true);
        var service = new AdvertisedService
        {
            ServiceId = "svc-http01", Name = "web", Transport = "tcp", Application = "http",
            PublishedPort = 8080, Path = "/app",
        };
        runtime.ApplyCatalog(2, "client-b", 9, 2, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), [service]);
        runtime.ApplyCatalog(2, "client-b", 9, 1, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), []);
        Assert.Single(runtime.RemoteServices());

        runtime.ApplyCatalog(2, "client-b", 9, 3, DateTimeOffset.UtcNow.ToString("O"), []);
        runtime.ApplyCatalog(2, "client-b", 9, 2, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), [service]);
        Assert.Empty(runtime.RemoteServices());
    }

    [Fact]
    public void ReconnectClearsCatalogRevisionHighWaterAndAcceptsCurrentSnapshot()
    {
        using var runtime = new PeerServiceRuntime(_ => { });
        runtime.SetRosterLookup(_ => new PeerServiceRuntime.RosterHint("100.96.0.2", true));
        runtime.ApplyConfig(Config(sharing: true, FreePort(), enabled: false));
        runtime.SetHasAuthorizedOnlinePeer(true);
        var service = new AdvertisedService
        {
            ServiceId = "svc-http01", Name = "web", Transport = "tcp", Application = "http",
            PublishedPort = 8080, Path = "/app",
        };
        runtime.ApplyCatalog(2, "client-b", 9, 7, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), [service]);
        Assert.Single(runtime.RemoteServices());

        runtime.SetHasAuthorizedOnlinePeer(false);
        Assert.Empty(runtime.RemoteServices());
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyCatalog(2, "client-b", 9, 7, DateTimeOffset.UtcNow.AddMinutes(1).ToString("O"), [service]);

        Assert.Single(runtime.RemoteServices());
    }

    [Fact]
    public void StableCatalogIsRenewedAcrossMultipleLeaseTtls()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyConfig(Config(sharing: true, port, enabled: true));
        WaitUntil(() => sent.Count > 0);
        sent.Clear();
        runtime.ProbeAndReport();
        Assert.Empty(sent);
        for (var elapsedTtls = 1; elapsedTtls <= 3; elapsedTtls++)
        {
            runtime.ForceReportRefreshForTest();
            runtime.ProbeAndReport();
            Assert.Equal(elapsedTtls, sent.Count);
        }
    }

    [Fact]
    public void ProbeTcpDetectsOpenAndClosedPorts()
    {
        Assert.False(PeerServiceDiscovery.IsLocalInterfaceTarget("10.255.255.254"));
        var port = FreePort();
        Assert.False(PeerServiceDiscovery.ProbeTcp("127.0.0.1", port, 200));
        using var listener = Listen(port);
        Assert.True(PeerServiceDiscovery.ProbeTcp("127.0.0.1", port, 400));
    }

    [Fact]
    public void LocalPauseStopsReportingWithoutChangingServerConfig()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyConfig(Config(sharing: true, port, enabled: true));
        WaitUntil(() => sent.Count > 0);
        sent.Clear();
        runtime.SetLocalPublished("svc-http01", false);
        WaitUntil(() => sent.Exists(item => item.Contains("\"enabled\":false") || item.Contains("\"services\":[]")));
        Assert.False(runtime.IsLocallyPublished("svc-http01"));
    }

    [Fact]
    public void TurningSharingOffWithdrawsPreviousReport()
    {
        var port = FreePort();
        using var listener = Listen(port);
        var sent = new List<string>();
        using var runtime = new PeerServiceRuntime(sent.Add);
        runtime.SetHasAuthorizedOnlinePeer(true);
        runtime.ApplyConfig(Config(sharing: true, port, enabled: true));
        WaitUntil(() => sent.Count > 0);
        sent.Clear();
        runtime.ApplyConfig(Config(sharing: false, port, enabled: true));
        WaitUntil(() => sent.Exists(item => item.Contains("\"enabled\":false")));
        Assert.Contains("service-report", sent[^1]);
    }

    [Fact]
    public void CollectedEnvironmentAdvertisesPeerServiceCapabilities()
    {
        var environment = ClientEnvironmentInfo.Collect(Microsoft.Extensions.Logging.Abstractions.NullLogger.Instance);
        Assert.Equal(2, environment.ClientPeerServiceCapabilities.Version);
        Assert.Equal(new[] { "http", "https", "ssh", "tcp", "udp" }, environment.ClientPeerServiceCapabilities.Applications);
    }

    [Fact]
    public async Task TcpBridgeEnforcesServerAuthoredSourceAcl()
    {
        using var target = Listen(0);
        var targetPort = ((IPEndPoint)target.LocalEndpoint).Port;
        var publishedPort = FreePort();
        var service = new LocalPeerService
        {
            ServiceId = "svc-acl01", Transport = "tcp", Application = "tcp",
            TargetHost = "127.0.0.1", TargetPort = targetPort, PublishedPort = publishedPort,
            AllowedPeerVirtualIps = ["127.0.0.2"],
        };
        using (var bridge = PeerServiceBridge.Bind("127.0.0.1", service, null))
        using (var caller = new TcpClient())
        {
            await caller.ConnectAsync(IPAddress.Loopback, publishedPort);
            using var timeout = new CancellationTokenSource(TimeSpan.FromMilliseconds(250));
            await Assert.ThrowsAnyAsync<OperationCanceledException>(async () =>
                await target.AcceptTcpClientAsync(timeout.Token));
        }

        publishedPort = FreePort();
        service.PublishedPort = publishedPort;
        service.AllowedPeerVirtualIps = ["127.0.0.1"];
        using var allowedBridge = PeerServiceBridge.Bind("127.0.0.1", service, null);
        using var allowedCaller = new TcpClient();
        await allowedCaller.ConnectAsync(IPAddress.Loopback, publishedPort);
        using var forwarded = await target.AcceptTcpClientAsync();
        Assert.True(forwarded.Connected);
    }

    private static PeerMeshConfig Config(bool sharing, int targetPort, bool enabled) => new()
    {
        Enabled = true,
        VirtualIp = "127.0.0.1",
        ServiceSharing = ServiceSharingStatus.Of(true, sharing, true),
        LocalServices =
        [
            new LocalPeerService
            {
                ServiceId = "svc-http01",
                Name = "web",
                Transport = "tcp",
                Application = "http",
                TargetHost = "127.0.0.1",
                TargetPort = targetPort,
                PublishedPort = 18080,
                Path = "/app",
                Enabled = enabled,
            },
        ],
    };

    private static int FreePort()
    {
        var listener = new TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }

    private static TcpListener Listen(int port)
    {
        var listener = new TcpListener(IPAddress.Loopback, port);
        listener.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
        listener.Start();
        return listener;
    }

    private static void WaitUntil(Func<bool> condition)
    {
        for (var i = 0; i < 50; i++)
        {
            if (condition())
            {
                return;
            }
            Thread.Sleep(20);
        }
        throw new InvalidOperationException("condition not met");
    }
}
